File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <bits/stdc++.h>
using namespace std;


int n;
int a[1000004], ret[1000004];
stack<int> st;
int main() {
	cin >> n;
	memset(ret, -1, sizeof(ret));
	for (int i=0; i <n; i++) {
		cin >> a[i];
		while (st.size() && a[st.top()] < a[i]) {
			ret[st.top()] = a[i];
			st.pop();
		}
		st.push(i);
		
	}

	for (int i = 0; i < n; i++) {
		cout << ret[i] << " ";
	}




}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
stack = []
ret = [-1 for _ in range(n)]

for i in range(n):
    # 스택이 비어있지 않고, 현재 값이 스택 top의 값보다 클 때
    while stack and arr[stack[-1]] < arr[i]:
        idx = stack.pop()
        ret[idx] = arr[i]
    
    # 현재 인덱스를 스택에 추가
    stack.append(i)

print(*ret)
